rust/otel-arrow-rust/src/otap/groups.rs

```diff
@@ -1431,144 +1430,386 @@ impl UnifiedDictionaryTypeSelector {
-                if dedup.len() > u16::MAX as usize {
-                    let selected_type = UnifiedDictionaryType::Native;
-                    self.selected_type = Some(selected_type.clone());
-                    return Ok(selected_type);
-                }
+fn bitmap_data_type_width(dt: &DataType) -> Option<usize> {
+    // FIXME: consider supporting 64-bit quantities including `Duration`, but that requires using
+    // `RoaringTreemap` instead of `RoaringBitmap`.
+    if let DataType::Dictionary(_key, value) = dt {
+        value.primitive_width()
+    } else {
+        dt.primitive_width()
+    }
+}
 
-                total_values_visited += 1;
-                let duplicates_visited = total_values_visited - dedup.len();
-                let possible_max_cardinality = self.total_batch_size - duplicates_visited;
-
-                // if we can determine that the values would fit in the desired key size, we can return early
-                if possible_max_cardinality < u16::MAX as usize
-                    && self.smallest_key_type == DataType::UInt16
-                {
-                    let selected_type = UnifiedDictionaryType::Dictionary(DataType::UInt16);
-                    self.selected_type = Some(selected_type.clone());
-                    return Ok(selected_type);
-                }
+/// I estimate cardinality for arrays of small elements (the size of a `u32` or smaller) using a
+/// `RoaringBitmap`. I support all the early termination features as my big sibling; the only
+/// difference is that I amortize early termination checking so that it doesn't happen for each
+/// element. This is because unlike a `HashSet` where `len()` is stored, computing cardinality for a
+/// `RoaringBitmap` takes some work; in addition, for tiny elements, the cost of `RoaringBitmap`
+/// insertion should be smaller than storing a slice in a `HashSet` so we don't mind doing some
+/// extra work there.
+struct SmallCardinalityEstimator<'parent> {
+    // these two are immutable inputs
+    arrays: &'parent [ArrayRef],
+    smallest_key_type: &'parent DataType,
+
+    // these are immutable after construction time
+    total_elements: usize, // Note: we only consider non-null items!
+    element_width_bytes: usize,
+
+    // and these are mutable state variables
+    elements_seen: usize,
+    bitmap: roaring::RoaringBitmap,
+}
+
+impl<'parent> SmallCardinalityEstimator<'parent> {
+    fn new(arrays: &'parent [ArrayRef], smallest_key_type: &'parent DataType) -> Option<Self> {
+        let total_elements: usize = arrays.iter().map(|v| v.len() - v.null_count()).sum();
+        let elements_seen = 0;
+        let bitmap = roaring::RoaringBitmap::new();
+
+        let element_width_bytes = bitmap_data_type_width(arrays.first()?.data_type())?;
+        // Because we're using RoaringBitmap, we're limited to `u32` and smaller types....
+        if element_width_bytes > 4 {
+            return None;
+        }
```

Comment:
Would it be slightly faster here to check the datatype before counting the valid elements? I guess this would only make a big difference if there were a large number of arrays